google-cloud-clients/google-cloud-spanner/src/main/java/com/google/cloud/spanner/SessionPool.java

```diff
@@ -206,7 +268,16 @@ public Timestamp getReadTimestamp() {
 
     @Override
     public TransactionContext begin() {
-      return delegate.begin();
+      while (true) {
+        try {
+          TransactionContext res = delegate.begin();
+          session.markUsed();
+          return res;
+        } catch (SessionNotFoundException e) {
+          session.maybeRecreateUnderlyingSession(e);
+          delegate = session.delegate.transactionManager();
```

Comment:
can you just use `return begin()` instead of adding the `while` loop?

Reply:
I replaced the while(true) loop with an internal method that is can be re-used in the SessionNotFound handler.